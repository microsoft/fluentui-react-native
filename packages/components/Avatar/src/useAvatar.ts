import { ImageProps, ImageSourcePropType } from 'react-native';
import { AvatarProps, AvatarInfo, AvatarState, AvatarColors, AvatarSize } from './Avatar.types';
import { PresenceBadgeProps } from '@fluentui-react-native/badge';
import { titles } from './titles';
import { getHashCodeWeb } from './getHashCode';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
/**
 * Re-usable hook for FURN Avatar.
 * This hook configures Avatar props and state for FURN Avatar.
 *
 * @param props user props sent to FURN Avatar
 * @returns configured props and state for FURN Avatar
 */
export const useAvatar = (props: AvatarProps): AvatarInfo => {
  const {
    avatarColor,
    active,
    accessibilityLabel,
    accessibilityRole,
    activeAppearance,
    badge,
    idForColor,
    image,
    imageUrl,
    initials,
    name,
    shape,
    transparentRing,
    icon,
    size,
    ...rest
  } = props;

  const showRing = active === 'active' && activeAppearance === 'ring';
  const showBadge = (!active || active === 'unset') && !!badge && !!badge.status;
  const accessibilityText = `${name || ''}${showBadge ? `, ${badge.status}` : ''}`;

  const badgeProps: PresenceBadgeProps = {
    size: 'small',
    ...badge,
  };

  const state: AvatarState = {
    showRing,
    transparentRing: !!transparentRing,
    showBadge,
  };

  let imageProps: ImageProps = {
    accessibilityLabel,
    source: imageUrl ? ({ uri: imageUrl } as ImageSourcePropType) : undefined,
  };

  if (image && image.source) {
    imageProps = {
      ...image,
      accessibilityLabel: image.accessibilityLabel ?? accessibilityLabel,
    };
  }

  const _initials = initials || getInitials(name);
  const avatarColorsIdx = getHashCodeWeb(idForColor ?? name ?? '') % AvatarColors.length;
  const _avatarColor = avatarColor === 'colorful' ? AvatarColors[avatarColorsIdx] : avatarColor;

  let iconProps = createIconProps(icon);
  const isFontIcon = !!(iconProps && iconProps.fontSource);
  if (isFontIcon) {
    const { fontSource, ...restIconProps } = iconProps;

    iconProps = {
      fontSource: {
        fontSize: fontSource.fontSize ?? getFontIconSize(size),
        ...iconProps.fontSource,
      },
      ...restIconProps,
    };
  }

  return {
    props: {
      accessible: true,
      accessibilityLabel: accessibilityLabel || accessibilityText,
      accessibilityRole: accessibilityRole ?? 'image',
      active,
      activeAppearance,
      avatarColor: _avatarColor,
      badge: badgeProps,
      badgeStatus: badge?.status,
      initials: _initials,
      icon: iconProps,
      image: imageProps,
      outOfOffice: badge?.outOfOffice,
      shape: shape ?? 'circular',
      size,
      ...rest,
    },
    state: {
      ...state,
    },
  };
};

/**
 * A function which takes a name and returns initials
 *
 * @param name
 * @returns initials. First letter of the word if only one word was provided.
 * First letters of the first and last words if more words were provided.
 * Words in braces, titles, special characters, parantheses and dashes should be ignored.
 */
export const getInitials = (name: string): string => {
  if (!name) {
    return '';
  }
  const ARABIC_ASIAN_REGEXP = new RegExp(
    '[\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\u1100-\u11FF\u3130-\u318F\uA960-\uA97F\uAC00-\uD7AF\uD7B0-\uD7FF\u3040-\u309F\u30A0-\u30FF\u3400-\u4DBF\u4E00-\u9FFF\uF900-\uFAFF\uD840-\uD869]',
    'g',
  );
  const words = removeRedundantCharacters(name);
  const wordsLength = words.length;
  const lastWordIdx = wordsLength - 1;
  const firstLetter = words[0].charAt(0).toUpperCase();
  const lastLetter = wordsLength > 1 ? words[lastWordIdx].charAt(0).toUpperCase() : '';
  const initials = `${firstLetter}${ARABIC_ASIAN_REGEXP.test(name) ? '' : lastLetter}`;
  return initials;
};

/**
 * A function which takes a name and returns array of valid words
 *
 * @param name
 * @returns array of words without phone numbers and special characters.
 */
export const removeRedundantCharacters = (name: string): string[] => {
  if (!name) {
    return [];
  }

  const MAX_NAME_LENGTH = 100;
  const _name = name.length > MAX_NAME_LENGTH ? name.trim().substring(0, MAX_NAME_LENGTH) : name;

  const WORDS_IN_BRACES_REGEXP = new RegExp('[(\\[\\{][^\\)\\]\\}]*[\\)\\]\\}]', 'g');
  const PHONE_NUMBER_REGEXP = new RegExp('(\\+|(\\d|\\s))', 'g');
  const SPECIAL_CHARACTERS_REGEXP = new RegExp('[!"#\'$%&*+,-./:;>=<?@^_`|~¡¢£¤¥¦§¨©ª«¬®ˉ°±²¼½¾¿×÷]', 'g');
  const words = _name
    .replace(WORDS_IN_BRACES_REGEXP, '')
    .replace(PHONE_NUMBER_REGEXP, ' ')
    .replace(SPECIAL_CHARACTERS_REGEXP, '')
    .trim()
    .split(' ');
  return removeTitlesFromName(words);
};

/**
 * A function which verifies whether words array contains title
 *
 * @param words array of words in name
 * @returns array of words without titles
 */
export const removeTitlesFromName = (words: string[]): string[] => {
  return words.filter((word) => !titles.has(word));
};

function getFontIconSize(size: AvatarSize) {
  if (size >= 25 && size <= 47) {
    return 20;
  } else if (size === 48) {
    return 24;
  } else if (size === 56) {
    return 28;
  } else if (size >= 57 && size <= 72) {
    return 32;
  } else if (size >= 73) {
    return 48;
  }
  return 16;
}
