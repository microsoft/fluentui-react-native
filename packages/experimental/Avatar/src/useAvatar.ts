import { AvatarProps, AvatarInfo, AvatarState } from './Avatar.types';
import { getPresenceIconSource } from './Avatar.helpers';

/**
 * Re-usable hook for FURN Avatar.
 * This hook configures Avatar props and state for FURN Avatar.
 *
 * @param props user props sent to FURN Avatar
 * @returns configured props and state for FURN Avatar
 */
export const useAvatar = (props: AvatarProps): AvatarInfo => {
  const { imageUrl, imageDescription, initials, presence, isOutOfOffice, ring, ...rest } = props;

  const personaPhotoSource =
    imageUrl === undefined
      ? undefined
      : {
          uri: imageUrl,
        };

  const iconSource = presence === undefined ? undefined : getPresenceIconSource(presence, isOutOfOffice || false);
  const showRing = !!ring;
  const transparentRing = !!ring?.transparent;

  const state: AvatarState = {
    iconSource,
    personaPhotoSource,
    showRing,
    transparentRing,
  };

  return {
    props: {
      children: initials,
      accessibilityLabel: imageDescription,
      ...rest,
    },
    state: {
      ...state,
    },
  };
};
