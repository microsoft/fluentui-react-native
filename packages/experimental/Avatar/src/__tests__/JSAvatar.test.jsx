import * as React from 'react';
import { JSAvatar } from '..';
import { getInitials, validateAlphabeticalCharacters, removeTitlesFromName } from '../useAvatar';
import * as renderer from 'react-test-renderer';

const alphabeticalTestNames = [
  'Marie',
  'Aadi KaPoor',
  'Marie Beaudouin',
  'Aadi Meni KaPoor',
  'Aadi Meni Ka Jr',
  'Will Little (Teams)',
  'Will Little-Tanaka',
  '-Aadi Kapoor',
  'Richard 23 Feynman',
  '* Richard *',
  'Richard Feynman [Teams]',
  'Richard Feynman {Teams}',
];
const alphabeticalInitials = ['M', 'AK', 'MB', 'AK', 'AK', 'WL', 'WL', 'AK', 'RF', 'R', 'RF', 'RF'];
const otherTestNames = ['(206) 123-4567', '耿盈盈']; // will return this case after bug bash - '耿 James'
const namesWithTitles = ['Mr Feynman', 'Dr Richard Feynman'];
const initialsForNameWithTitles = ['F', 'RF'];

it('renders Avatar', () => {
  const tree = renderer.create(<JSAvatar />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('validateAlphabeticalCharacters method', () => {
  it('validates alphabetical characters', () => {
    for (let i = 0; i < alphabeticalTestNames.length; i++) {
      expect(validateAlphabeticalCharacters(alphabeticalTestNames[i])).toBeTruthy();
    }
    for (let i = 0; i < otherTestNames.length; i++) {
      expect(validateAlphabeticalCharacters(otherTestNames[i])).toBeFalsy();
    }
  });

  it("doesn't validate undefined name", () => {
    expect(validateAlphabeticalCharacters(undefined)).toBeFalsy();
  });
  it("doesn't validate an empty string", () => {
    expect(validateAlphabeticalCharacters('')).toBeFalsy();
  });
  it("doesn't validate null name", () => {
    expect(validateAlphabeticalCharacters(null)).toBeFalsy();
  });
});

describe('removeTitlesFromName method', () => {
  it('removes titles from the name', () => {
    const words = ['Mr', 'Faynman'];
    expect(removeTitlesFromName(words)).toStrictEqual(['Faynman']);
  });
  // it('removes multiword titles from the name', () => {
  //   const name = 'Consul General Richard Faynman';
  //   expect(removeTitlesFromName(name)).toStrictEqual(['Richard Faynman']);
  // });
});

describe('getInitials method', () => {
  it('returns correct initials', () => {
    const test = getInitials('-Test   str  with *spaces');
    expect(test).toBe('TS');
  });

  it('returns an empty string if name is undefined', () => {
    expect(getInitials(undefined)).toBe('');
  });

  it('return initials for alphabetical words', () => {
    for (let i = 0; i < alphabeticalTestNames.length; i++) {
      expect(getInitials(alphabeticalTestNames[i])).toBe(alphabeticalInitials[i]);
    }
  });

  it('return initials for words with titles', () => {
    for (let i = 0; i < namesWithTitles.length; i++) {
      expect(getInitials(namesWithTitles[i])).toBe(initialsForNameWithTitles[i]);
    }
  });

  it("returns initials when there's a coma", () => {
    expect(getInitials('Second, First')).toEqual('SF');
  });

  // Unit tests from Web
  it('handles null inputs', () => {
    expect(getInitials(null)).toEqual('');
    expect(getInitials(undefined)).toEqual('');
  });

  // it('calculates an expected initials in LTR for non-ASCII characters', () => {
  //   expect(getInitials('Írissa Þórðardóttir')).toEqual('ÍÞ');
  //   expect(getInitials('Øyvind Åsen')).toEqual('ØÅ');
  // });

  it('calculates an expected initials in LTR with a hypen', () => {
    expect(getInitials('David Zearing-Goff')).toEqual('DZ');
  });

  // it('calculates an expected initials in LTR with numbers', () => {
  //   expect(getInitials('4lex 5loo')).toEqual('45');
  // });
  it('calculates an expected initials in LTR with multiple parentheses, extra spaces, and unwanted characters', () => {
    const result = getInitials(' !@#$%^&*()=+ (Alpha) David   (The man) `~<>,./?[]{}|   Goff   (Gamma)    ');
    expect(result).toEqual('DG');
  });

  it('calculates an expected initials in LTR with multiple types of unwanted text', () => {
    const result = getInitials(' !@#$%^&*()=+ (Alpha) David   (The man) `~<>,./?[]{}|   Goff   (Gamma)  [Beta]  ');
    expect(result).toEqual('DG');
  });

  it('calculates an expected initials for Arabic names', () => {
    expect(getInitials('خسرو رحیمی')).toEqual('');
  });

  it('calculates an expected initials for Chinese names', () => {
    expect(getInitials('桂英')).toEqual('');
    expect(getInitials('佳')).toEqual('');
    expect(getInitials('宋智洋')).toEqual('');
  });

  it('calculates an expected initials for Korean names', () => {
    expect(getInitials('강현')).toEqual('');
    expect(getInitials('최종래')).toEqual('');
    expect(getInitials('남궁 성종')).toEqual('');
  });

  it('calculates an expected initials for Japanese names', () => {
    expect(getInitials('松田')).toEqual('');
    expect(getInitials('海野')).toEqual('');
    expect(getInitials('かり')).toEqual('');
  });

  // it('validates phone numbers', () => {
  // expect(getInitials('12345678')).toEqual('');
  // expect(getInitials('+1 (555) 123-4567 ext.4567')).toEqual('');
  // expect(getInitials('+47 12 34 56 78 (X 5678)')).toEqual('4');
  // expect(getInitials('47 12 34')).toEqual('');
  // expect(getInitials('47 12')).toEqual('');
  // expect(getInitials('1 Ext 2')).toEqual('');
  // result = getInitials('James Ext 2');
  // expect(result).toEqual('JE');
  // result = getInitials('1x1');
  // expect(result).toEqual('');
  // result = getInitials('1y1');
  // expect(result).toEqual('');
  // result = getInitials('1');
  // expect(result).toEqual('');
  // result = getInitials('A 2');
  // expect(result).toEqual('A');
  // });
});
