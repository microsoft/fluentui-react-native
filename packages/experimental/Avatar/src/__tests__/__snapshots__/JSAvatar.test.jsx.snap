// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar rendering renders Avatar 1`] = `
<View
  accessibilityLabel=""
  accessibilityRole="image"
  accessible={true}
  icon={null}
  initials=""
  shape="circular"
  style={
    Object {
      "alignItems": "center",
      "height": 24,
      "justifyContent": "center",
      "opacity": 1,
      "width": 24,
    }
  }
>
  <View
    accessible={false}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#e6e6e6",
        "borderColor": "#ffffff",
        "borderRadius": 9999,
        "borderWidth": 0,
        "height": 24,
        "justifyContent": "center",
        "marginLeft": 2,
        "marginTop": 2,
        "width": 24,
      }
    }
  >
    <View
      accessibilityRole="image"
      accessible={false}
      style={
        Object {
          "fontSize": 16,
          "height": 16,
          "position": "absolute",
          "width": 16,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight="100%"
        bbWidth="100%"
        focusable={false}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": "100%",
              "width": "100%",
            },
          ]
        }
        vbHeight={16}
        vbWidth={14}
      >
        <RNSVGGroup>
          <RNSVGPath
            d="M7 0C4.79086 0 3 1.79086 3 4C3 6.20914 4.79086 8 7 8C9.20914 8 11 6.20914 11 4C11 1.79086 9.20914 0 7 0ZM4 4C4 2.34315 5.34315 1 7 1C8.65685 1 10 2.34315 10 4C10 5.65685 8.65685 7 7 7C5.34315 7 4 5.65685 4 4ZM2.00873 9C0.903151 9 0 9.88687 0 11C0 12.6912 0.83281 13.9663 2.13499 14.7966C3.41697 15.614 5.14526 16 7 16C8.85474 16 10.583 15.614 11.865 14.7966C13.1672 13.9663 14 12.6912 14 11C14 9.89557 13.1045 9.00001 12 9.00001L2.00873 9ZM1 11C1 10.4467 1.44786 10 2.00873 10L12 10C12.5522 10 13 10.4478 13 11C13 12.3088 12.3777 13.2837 11.3274 13.9534C10.2568 14.636 8.73511 15 7 15C5.26489 15 3.74318 14.636 2.67262 13.9534C1.62226 13.2837 1 12.3088 1 11Z"
            fill={
              Array [
                2,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
</View>
`;
